app/renderer/renderer.py

```diff
@@ -112,47 +136,53 @@ def _render_navigation_meta(self):
 
         return navigation_meta
 
-    def _augment_responses(self):
-        # Augment the schema with user responses and validation results
-        all_responses = self._response_store.get_responses()
-        for item_id, value in all_responses.items():
-            item = self._schema.get_item_by_id(item_id)
-            if item:
-                item.value = value
-                validation_result = self._validation_store.get_result(item_id)
+    def _augment_response(self, response):
+        if response.id in self._response_store.get_responses().keys():
+            value = self._response_store.get_response(response.id)
+            if value is not None:
+                response.value = value
+                validation_result = self._validation_store.get_result(response.id)
                 if validation_result:
-                    item.is_valid = validation_result.is_valid
-                    item.errors = validation_result.get_errors()
-                    item.warnings = validation_result.get_warnings()
+                    response.is_valid = validation_result.is_valid
+                    response.errors = validation_result.get_errors()
+                    response.warnings = validation_result.get_warnings()
                 else:
-                    item.is_valid = None
-                    item.errors = None
-                    item.warnings = None
+                    response.is_valid = None
+                    response.errors = None
+                    response.warnings = None
 
     def _augment_questionnaire(self):
         errors = OrderedDict()
         warnings = OrderedDict()
 
         # loops through the Schema and get errors and warnings in order
+        # augments each item in the schema as required
         for group in self._schema.groups:
+            self._plumber.plumb_item(group)
```

Comment:
As a minimum these steps (lines 161-166) can be factored out onto a function and called for each of the repeats in the nested loops below. Further refactoring may be possible beyond this to clean this up.